### ISSUE TITLE
Issues with UNIQUE and ON CONFLICT after updating to 3.17.3 (SQLite)

### ISSUE THREAD

**Issue by Jsmith7861** [jOOQ/jOOQ#13879]
Hello,

My organization recently upgraded to jOOQ `3.17.3` from V `3.14.15`. After upgrading, we have started to have issues with code generation. We currently use jOOQ to generate code from a SQLite database. Previously, `includeUniqueKeys`, `includeForeignKeys`, and `includePrimaryKeys` were all enabled without issue. After updating, we now have parsing errors for the same schema.

Below, I have listed a few simplified examples of various SQLite tables that are causing issue in jOOQ `3.17.3` and were previously fine in V `3.14.15`.

PASSING:
`CREATE TABLE application (ID TEXT NOT NULL PRIMARY KEY, Name TEXT UNIQUE NOT NULL);`
`CREATE TABLE application (ID TEXT NOT NULL PRIMARY KEY UNIQUE, Name TEXT NOT NULL);`

FAILING:
`CREATE TABLE application (ID TEXT NOT NULL PRIMARY KEY UNIQUE, Name TEXT UNIQUE NOT NULL);`

```
14:49:52 SEVERE An exception occurred while parsing a DDL script: Token ')' expected: [1:56] ...E TABLE application (ID TEXT NOT NULL PRIMARY KEY [*]UNIQUE, Name TEXT UNIQUE NOT NULL). Please report this error to https://github.com/jOOQ/jOOQ/issues/new
org.jooq.impl.ParserException: Token ')' expected: [1:56] ...E TABLE application (ID TEXT NOT NULL PRIMARY KEY [*]UNIQUE, Name TEXT UNIQUE NOT NULL)
        at org.jooq.impl.DefaultParseContext.expected(ParserImpl.java:14095)
        at org.jooq.impl.DefaultParseContext.parse(ParserImpl.java:13352)
        at org.jooq.impl.DefaultParseContext.parse(ParserImpl.java:13347)
        at org.jooq.impl.DefaultParseContext.parseCreateTable(ParserImpl.java:4417)
        at org.jooq.impl.DefaultParseContext.parseCreate(ParserImpl.java:2846)
        at org.jooq.impl.DefaultParseContext.parseQuery(ParserImpl.java:1097)
        at org.jooq.impl.DefaultParseContext.lambda$parse$0(ParserImpl.java:877)
        at org.jooq.impl.DefaultParseContext.wrap(ParserImpl.java:14322)
        at org.jooq.impl.DefaultParseContext.parse(ParserImpl.java:865)
        at org.jooq.impl.ParserImpl.parse(ParserImpl.java:743)
        at org.jooq.impl.ParserImpl.parse(ParserImpl.java:738)
        at org.jooq.impl.InterpreterMetaProvider.loadSource(InterpreterMetaProvider.java:106)
        at org.jooq.impl.InterpreterMetaProvider.provide(InterpreterMetaProvider.java:92)
        at org.jooq.impl.SourceMetaProvider.provide(SourceMetaProvider.java:80)
        at org.jooq.impl.DefaultDSLContext.meta(DefaultDSLContext.java:462)
        at org.jooq.impl.MetaImpl$MetaSchema.lambda$initUksSQLite$5(MetaImpl.java:517)
        at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:721)
        at org.jooq.impl.MetaImpl$MetaSchema.initUksSQLite(MetaImpl.java:504)
        at org.jooq.impl.MetaImpl$MetaSchema.getUks(MetaImpl.java:450)
        at org.jooq.impl.MetaImpl$MetaSchema.lambda$getTables$2(MetaImpl.java:434)
        at org.jooq.impl.Tools.map(Tools.java:2313)
        at org.jooq.impl.MetaImpl$MetaSchema.getTables(MetaImpl.java:406)
        at org.jooq.impl.Snapshot$SnapshotSchema.<init>(Snapshot.java:125)
        at org.jooq.impl.Snapshot$SnapshotCatalog.lambda$new$0(Snapshot.java:100)
        at org.jooq.impl.Tools.map(Tools.java:2313)
        at org.jooq.impl.Snapshot$SnapshotCatalog.<init>(Snapshot.java:100)
        at org.jooq.impl.Snapshot.lambda$getCatalogs0$0(Snapshot.java:92)
        at org.jooq.impl.Tools.map(Tools.java:2313)
        at org.jooq.impl.Snapshot.getCatalogs0(Snapshot.java:92)
        at org.jooq.impl.AbstractMeta.initCatalogs(AbstractMeta.java:124)
        at org.jooq.impl.AbstractMeta.getCatalogs(AbstractMeta.java:116)
        at org.jooq.impl.Snapshot.<init>(Snapshot.java:80)
        at org.jooq.impl.AbstractMeta.snapshot(AbstractMeta.java:537)
        at org.jooq.meta.sqlite.SQLiteDatabase.snapshot(SQLiteDatabase.java:469)
        at org.jooq.meta.sqlite.SQLiteDatabase.loadPrimaryKeys(SQLiteDatabase.java:236)
        at org.jooq.meta.AbstractDatabase.lambda$getRelations0$25(AbstractDatabase.java:3109)
        at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:3749)
        at org.jooq.meta.AbstractDatabase.getRelations0(AbstractDatabase.java:3109)
        at org.jooq.meta.AbstractDatabase.lambda$getRelations$18(AbstractDatabase.java:2787)
        at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:3749)
        at org.jooq.meta.AbstractDatabase.getRelations(AbstractDatabase.java:2787)
        at org.jooq.meta.DefaultColumnDefinition.getPrimaryKey(DefaultColumnDefinition.java:154)
        at org.jooq.meta.AbstractTableDefinition.getPrimaryKey(AbstractTableDefinition.java:96)
        at org.jooq.codegen.JavaGenerator.generateTable(JavaGenerator.java:5872)
        at org.jooq.codegen.JavaGenerator.generateTables(JavaGenerator.java:5851)
        at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:652)
        at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:607)
        at org.jooq.codegen.JavaGenerator.generate0(JavaGenerator.java:519)
        at org.jooq.codegen.AbstractGenerator.generate(AbstractGenerator.java:187)
        at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:227)
        at org.jooq.codegen.GenerationTool.run0(GenerationTool.java:942)
        at org.jooq.codegen.GenerationTool.run(GenerationTool.java:240)
        at org.jooq.codegen.GenerationTool.generate(GenerationTool.java:235)
        at org.jooq.codegen.GenerationTool.main(GenerationTool.java:207)
```

PASSING:
`CREATE TABLE application (ID TEXT NOT NULL PRIMARY KEY, Name TEXT NOT NULL ON CONFLICT REPLACE);`

FAILING:
`CREATE TABLE application (ID TEXT NOT NULL PRIMARY KEY, Name TEXT NOT NULL UNIQUE ON CONFLICT REPLACE);`

```
14:51:51 SEVERE An exception occurred while parsing a DDL script: Token ')' expected: [1:83] ...T NOT NULL PRIMARY KEY, Name TEXT NOT NULL UNIQUE [*]ON CONFLICT REPLACE). Please report this error to https://github.com/jOOQ/jOOQ/issues/new
org.jooq.impl.ParserException: Token ')' expected: [1:83] ...T NOT NULL PRIMARY KEY, Name TEXT NOT NULL UNIQUE [*]ON CONFLICT REPLACE)
        at org.jooq.impl.DefaultParseContext.expected(ParserImpl.java:14095)
        at org.jooq.impl.DefaultParseContext.parse(ParserImpl.java:13352)
        at org.jooq.impl.DefaultParseContext.parse(ParserImpl.java:13347)
        at org.jooq.impl.DefaultParseContext.parseCreateTable(ParserImpl.java:4417)
        at org.jooq.impl.DefaultParseContext.parseCreate(ParserImpl.java:2846)
        at org.jooq.impl.DefaultParseContext.parseQuery(ParserImpl.java:1097)
        at org.jooq.impl.DefaultParseContext.lambda$parse$0(ParserImpl.java:877)
        at org.jooq.impl.DefaultParseContext.wrap(ParserImpl.java:14322)
        at org.jooq.impl.DefaultParseContext.parse(ParserImpl.java:865)
        at org.jooq.impl.ParserImpl.parse(ParserImpl.java:743)
        at org.jooq.impl.ParserImpl.parse(ParserImpl.java:738)
        at org.jooq.impl.InterpreterMetaProvider.loadSource(InterpreterMetaProvider.java:106)
        at org.jooq.impl.InterpreterMetaProvider.provide(InterpreterMetaProvider.java:92)
        at org.jooq.impl.SourceMetaProvider.provide(SourceMetaProvider.java:80)
        at org.jooq.impl.DefaultDSLContext.meta(DefaultDSLContext.java:462)
        at org.jooq.impl.MetaImpl$MetaSchema.lambda$initUksSQLite$5(MetaImpl.java:517)
        at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:721)
        at org.jooq.impl.MetaImpl$MetaSchema.initUksSQLite(MetaImpl.java:504)
        at org.jooq.impl.MetaImpl$MetaSchema.getUks(MetaImpl.java:450)
        at org.jooq.impl.MetaImpl$MetaSchema.lambda$getTables$2(MetaImpl.java:434)
        at org.jooq.impl.Tools.map(Tools.java:2313)
        at org.jooq.impl.MetaImpl$MetaSchema.getTables(MetaImpl.java:406)
        at org.jooq.impl.Snapshot$SnapshotSchema.<init>(Snapshot.java:125)
        at org.jooq.impl.Snapshot$SnapshotCatalog.lambda$new$0(Snapshot.java:100)
        at org.jooq.impl.Tools.map(Tools.java:2313)
        at org.jooq.impl.Snapshot$SnapshotCatalog.<init>(Snapshot.java:100)
        at org.jooq.impl.Snapshot.lambda$getCatalogs0$0(Snapshot.java:92)
        at org.jooq.impl.Tools.map(Tools.java:2313)
        at org.jooq.impl.Snapshot.getCatalogs0(Snapshot.java:92)
        at org.jooq.impl.AbstractMeta.initCatalogs(AbstractMeta.java:124)
        at org.jooq.impl.AbstractMeta.getCatalogs(AbstractMeta.java:116)
        at org.jooq.impl.Snapshot.<init>(Snapshot.java:80)
        at org.jooq.impl.AbstractMeta.snapshot(AbstractMeta.java:537)
        at org.jooq.meta.sqlite.SQLiteDatabase.snapshot(SQLiteDatabase.java:469)
        at org.jooq.meta.sqlite.SQLiteDatabase.loadPrimaryKeys(SQLiteDatabase.java:236)
        at org.jooq.meta.AbstractDatabase.lambda$getRelations0$25(AbstractDatabase.java:3109)
        at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:3749)
        at org.jooq.meta.AbstractDatabase.getRelations0(AbstractDatabase.java:3109)
        at org.jooq.meta.AbstractDatabase.lambda$getRelations$18(AbstractDatabase.java:2787)
        at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:3749)
        at org.jooq.meta.AbstractDatabase.getRelations(AbstractDatabase.java:2787)
        at org.jooq.meta.DefaultColumnDefinition.getPrimaryKey(DefaultColumnDefinition.java:154)
        at org.jooq.meta.AbstractTableDefinition.getPrimaryKey(AbstractTableDefinition.java:96)
        at org.jooq.codegen.JavaGenerator.generateTable(JavaGenerator.java:5872)
        at org.jooq.codegen.JavaGenerator.generateTables(JavaGenerator.java:5851)
        at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:652)
        at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:607)
        at org.jooq.codegen.JavaGenerator.generate0(JavaGenerator.java:519)
        at org.jooq.codegen.AbstractGenerator.generate(AbstractGenerator.java:187)
        at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:227)
        at org.jooq.codegen.GenerationTool.run0(GenerationTool.java:942)
        at org.jooq.codegen.GenerationTool.run(GenerationTool.java:240)
        at org.jooq.codegen.GenerationTool.generate(GenerationTool.java:235)
        at org.jooq.codegen.GenerationTool.main(GenerationTool.java:207)
```

We are also now having issues with our forced type, which you can see the configuration of below. This was not an issue on the previous version either:

```
database {
                        name = 'org.jooq.meta.sqlite.SQLiteDatabase'
                        includes = '.*'
                        forcedTypes {
                            forcedType {
                                name = 'UUID'
                                includeExpression = '.*'
                                includeTypes = 'GUID|GUID\\s\\(32\\)'
                            }
                        }
                }
```

PASSING:
`CREATE TABLE application (ID GUID(32) NOT NULL PRIMARY KEY UNIQUE, Name TEXT NOT NULL);`

FAILING:
`CREATE TABLE application (ID GUID(32) NOT NULL PRIMARY KEY, Name TEXT NOT NULL UNIQUE);`

```
16:37:10 SEVERE An exception occurred while parsing a DDL script: Token ')' expected: [1:34] CREATE TABLE application (ID GUID[*](32) NOT NULL PRIMARY KEY, Name text NOT NULL UNIQUE). Please report this error to https://github.com/jOOQ/jOOQ/issues/new
org.jooq.impl.ParserException: Token ')' expected: [1:34] CREATE TABLE application (ID GUID[*](32) NOT NULL PRIMARY KEY, Name text NOT NULL UNIQUE)
        at org.jooq.impl.DefaultParseContext.expected(ParserImpl.java:14095)
        at org.jooq.impl.DefaultParseContext.parse(ParserImpl.java:13352)
        at org.jooq.impl.DefaultParseContext.parse(ParserImpl.java:13347)
        at org.jooq.impl.DefaultParseContext.parseCreateTable(ParserImpl.java:4417)
        at org.jooq.impl.DefaultParseContext.parseCreate(ParserImpl.java:2846)
        at org.jooq.impl.DefaultParseContext.parseQuery(ParserImpl.java:1097)
        at org.jooq.impl.DefaultParseContext.lambda$parse$0(ParserImpl.java:877)
        at org.jooq.impl.DefaultParseContext.wrap(ParserImpl.java:14322)
        at org.jooq.impl.DefaultParseContext.parse(ParserImpl.java:865)
        at org.jooq.impl.ParserImpl.parse(ParserImpl.java:743)
        at org.jooq.impl.ParserImpl.parse(ParserImpl.java:738)
        at org.jooq.impl.InterpreterMetaProvider.loadSource(InterpreterMetaProvider.java:106)
        at org.jooq.impl.InterpreterMetaProvider.provide(InterpreterMetaProvider.java:92)
        at org.jooq.impl.SourceMetaProvider.provide(SourceMetaProvider.java:80)
        at org.jooq.impl.DefaultDSLContext.meta(DefaultDSLContext.java:462)
        at org.jooq.impl.MetaImpl$MetaSchema.lambda$initUksSQLite$5(MetaImpl.java:517)
        at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:721)
        at org.jooq.impl.MetaImpl$MetaSchema.initUksSQLite(MetaImpl.java:504)
        at org.jooq.impl.MetaImpl$MetaSchema.getUks(MetaImpl.java:450)
        at org.jooq.impl.MetaImpl$MetaSchema.lambda$getTables$2(MetaImpl.java:434)
        at org.jooq.impl.Tools.map(Tools.java:2313)
        at org.jooq.impl.MetaImpl$MetaSchema.getTables(MetaImpl.java:406)
        at org.jooq.impl.Snapshot$SnapshotSchema.<init>(Snapshot.java:125)
        at org.jooq.impl.Snapshot$SnapshotCatalog.lambda$new$0(Snapshot.java:100)
        at org.jooq.impl.Tools.map(Tools.java:2313)
        at org.jooq.impl.Snapshot$SnapshotCatalog.<init>(Snapshot.java:100)
        at org.jooq.impl.Snapshot.lambda$getCatalogs0$0(Snapshot.java:92)
        at org.jooq.impl.Tools.map(Tools.java:2313)
        at org.jooq.impl.Snapshot.getCatalogs0(Snapshot.java:92)
        at org.jooq.impl.AbstractMeta.initCatalogs(AbstractMeta.java:124)
        at org.jooq.impl.AbstractMeta.getCatalogs(AbstractMeta.java:116)
        at org.jooq.impl.Snapshot.<init>(Snapshot.java:80)
        at org.jooq.impl.AbstractMeta.snapshot(AbstractMeta.java:537)
        at org.jooq.meta.sqlite.SQLiteDatabase.snapshot(SQLiteDatabase.java:469)
        at org.jooq.meta.sqlite.SQLiteDatabase.loadPrimaryKeys(SQLiteDatabase.java:236)
        at org.jooq.meta.AbstractDatabase.lambda$getRelations0$25(AbstractDatabase.java:3109)
        at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:3749)
        at org.jooq.meta.AbstractDatabase.getRelations0(AbstractDatabase.java:3109)
        at org.jooq.meta.AbstractDatabase.lambda$getRelations$18(AbstractDatabase.java:2787)
        at org.jooq.meta.AbstractDatabase.onError(AbstractDatabase.java:3749)
        at org.jooq.meta.AbstractDatabase.getRelations(AbstractDatabase.java:2787)
        at org.jooq.meta.DefaultColumnDefinition.getPrimaryKey(DefaultColumnDefinition.java:154)
        at org.jooq.meta.AbstractTableDefinition.getPrimaryKey(AbstractTableDefinition.java:96)
        at org.jooq.codegen.JavaGenerator.generateTable(JavaGenerator.java:5872)
        at org.jooq.codegen.JavaGenerator.generateTables(JavaGenerator.java:5851)
        at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:652)
        at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:607)
        at org.jooq.codegen.JavaGenerator.generate0(JavaGenerator.java:519)
        at org.jooq.codegen.AbstractGenerator.generate(AbstractGenerator.java:187)
        at org.jooq.codegen.JavaGenerator.generate(JavaGenerator.java:227)
        at org.jooq.codegen.GenerationTool.run0(GenerationTool.java:942)
        at org.jooq.codegen.GenerationTool.run(GenerationTool.java:240)
        at org.jooq.codegen.GenerationTool.generate(GenerationTool.java:235)
        at org.jooq.codegen.GenerationTool.main(GenerationTool.java:207)
```

We are trying to figure out what exactly is going on, but aren't entirely sure. It seems as if something is going on with the `UNIQUE` and `ON CONFLICT` statements. Please let me know if I can be of assistance, or if there is some sort of workaround.

Thank you!

**Comment by lukaseder** [comment]
Thanks a lot for your report.

To be clear, I think these errors shouldn't affect code generation per se, they only limit usage of some features on affected tables, e.g. the constraints couldn't be reverse engineered? The stack trace is logged mostly to help users troubleshoot missing constraints as well as to be able to better report such issues.

The change that likely produced this problem on your end was https://github.com/jOOQ/jOOQ/issues/11286. SQLite doesn't offer very good dictionary views to reverse engineer the schema. The various `pragma` commands are severely limited, which is why jOOQ eventually started parsing the DDL instead.

You've mixed a few issues, I'll try to address them individually via separate issues here:

### `PRIMARY KEY UNIQUE`

I'm not sure what the purpose of adding `UNIQUE` to a `PRIMARY KEY` is? The workaround here is to simply remove `UNIQUE`, which is redundant. But obviously, if SQL dialects parse this, then so must jOOQ's parser. I've created a separate issue for this:
https://github.com/jOOQ/jOOQ/issues/13880

### `ON CONFLICT`

We should probably parse and currently ignore multiple of these additional DDL clauses, including `ON CONFLICT`: https://github.com/jOOQ/jOOQ/issues/13881

### `GUID(32)`

SQLite doesn't really have such a data type, it just translates to affinity `NONE`, so are you mostly using this for documentation purposes? See: https://stackoverflow.com/q/18954130/521799

I've created a new issue for this:
https://github.com/jOOQ/jOOQ/issues/13882

**Comment by lukaseder** [comment]
I'm closing this issue here as a duplicate of the other 3 created. Individual problems are usually addressed separately under separate github issues, which simplifies discussion and tracking. It is a single problem for you, but for jOOQ maintenance, the 3 problems are unrelated.

Thanks again for your report!

**Comment by lukaseder** [comment]
All 3 issues have been implemented in jOOQ 3.18.0, 3.17.4, and 3.16.10

**Comment by Jsmith7861** [comment]
Lukas,

Thanks for your quick reply! I apologize for the lengthy report (with multiple issues). Much of our DB has existed for a while, and I had similar comments/concerns as yours here. Not much I can do about them, but I do appreciate your concerns here!

Thanks again!
